erigon-lib/downloader/mdbx_piece_completion.go

```diff
@@ -89,59 +89,54 @@ func (m *mdbxPieceCompletion) Get(pk metainfo.PieceKey) (cn storage.Completion,
 	return
 }
 
-func (m *mdbxPieceCompletion) Set(pk metainfo.PieceKey, b bool) error {
+func (m *mdbxPieceCompletion) Set(pk metainfo.PieceKey, b bool, awaitFlush bool) error {
 	if c, err := m.Get(pk); err == nil && c.Ok && c.Complete == b {
 		return nil
 	}
 
-	m.mu.Lock()
-	defer m.mu.Unlock()
-
-	var tx kv.RwTx
-	var err error
-	// On power-off recent "no-sync" txs may be lost.
-	// It will cause 2 cases of in-consistency between files on disk and db metadata:
-	//  - Good piece on disk and recent "complete"   db marker lost. Self-Heal by re-download.
-	//  - Bad  piece on disk and recent "incomplete" db marker lost. No Self-Heal. Means: can't afford loosing recent "incomplete" markers.
-	// FYI: Fsync of torrent pieces happenng before storing db markers: https://github.com/anacrolix/torrent/blob/master/torrent.go#L2026
-	//
-	// Mainnet stats:
-	//  call amount 2 minutes complete=100K vs incomple=1K
-	//  1K fsyncs/2minutes it's quite expensive, but even on cloud (high latency) drive it allow download 100mb/s
-	//  and Erigon doesn't do anything when downloading snapshots
-	if b {
-		completed, ok := m.completed[pk.InfoHash]
-
-		if !ok {
-			completed = &roaring.Bitmap{}
-			m.completed[pk.InfoHash] = completed
-		}
-
-		completed.Add(uint32(pk.Index))
-
-		if flushed, ok := m.flushed[pk.InfoHash]; !ok || !flushed.Contains(uint32(pk.Index)) {
-			return nil
+	persist := func() bool {
+		m.mu.Lock()
+		defer m.mu.Unlock()
+
+		if b {
+			completed, ok := m.completed[pk.InfoHash]
+
+			if !ok {
+				completed = &roaring.Bitmap{}
+				m.completed[pk.InfoHash] = completed
+			}
+
+			completed.Add(uint32(pk.Index))
+
+			// when files are being downloaded flushing is asynchronous so we want to wait for
+			// the confirm before committing to the DB.  This means that if the program is
+			// abnormally terminated the piece will be re-downloaded
+			if awaitFlush {
+				if flushed, ok := m.flushed[pk.InfoHash]; !ok || !flushed.Contains(uint32(pk.Index)) {
+					return false
+				}
+			}
+		} else {
+			if completed, ok := m.completed[pk.InfoHash]; ok {
+				completed.Remove(uint32(pk.Index))
+			}
 		}
-	} else {
-		if completed, ok := m.completed[pk.InfoHash]; ok {
-			completed.Remove(uint32(pk.Index))
-		}
-	}
+		return true
+	}()
 
-	tx, err = m.db.BeginRw(context.Background())
-	if err != nil {
-		return err
+	if !persist {
+		return nil
 	}
 
-	defer tx.Rollback()
-
-	err = putCompletion(tx, pk.InfoHash, uint32(pk.Index), b)
-
-	if err != nil {
-		return err
+	if awaitFlush {
```

Comment:
Only comment: i don't understand why `if awaitFlush == true` then Update, else Batch. Why not opposite?

Reply:
One more thing: if `b == false` maybe you need use inversed logic:
- mark in DB immediately as `false` - not waiting for `Flush`
- then on restart - you will not loose that fact - that piece was marked as bad 

In another words: save in db `true` only after flush (after ensure that data will be good after restart), save in db `false` ASAP (to be conservative. to re-validate this piece after restart: even piece is good or bad on disk).